contracts/src/L2/L2EjectionController.sol

```diff
@@ -0,0 +1,110 @@
+    }
+
+    /**
+     * @dev Called by the L2 eth registry when a user ejects a name to L1.
+     *
+     * @param tokenId The token ID of the name being ejected
+     * @param l1Owner The address that will own the name on L1
+     * @param l1Subregistry The subregistry address to use on L1
+     * @param l1Resolver The resolver address to use on L1
+     */
+    function ejectToL1(uint256 tokenId, address l1Owner, address l1Subregistry, address l1Resolver) external {
+        if (registry.ownerOf(tokenId) != address(this)) {
+            revert NotTokenOwner(tokenId);
+        }
+
+        uint64 expiry = registry.getExpiry(tokenId);
+
+        registry.setSubregistry(tokenId, IRegistry(address(0)));
+
+        // bridge will listen for this event        
+        emit NameEjectedToL1(tokenId, l1Owner, l1Subregistry, l1Resolver, expiry);
+    }
+
+    /**
+     * @dev Called by the cross-chain messaging system when a name is being migrated back to L2.
+     *
+     * @param tokenId The token ID of the name being migrated
+     * @param l2Owner The address that will own the name on L2
+     * @param l2Subregistry The subregistry address to use on L2
+     * @param l2Resolver The resolver address to use on L2
+     */
+    function completeMigrationToL2(
+        uint256 tokenId,
+        address l2Owner,
+        address l2Subregistry,
+        address l2Resolver
+    ) external {
+        if (registry.ownerOf(tokenId) != address(this)) {
+            revert NotTokenOwner(tokenId);
+        }
+
+        registry.setSubregistry(tokenId, IRegistry(l2Subregistry));
+        registry.setResolver(tokenId, l2Resolver);
+        registry.safeTransferFrom(address(this), l2Owner, tokenId, 1, "");
+
+        emit NameMigratedToL2(tokenId, l2Owner, l2Subregistry, l2Resolver);
+    }
+
+    /**
+     * Implements ERC165.supportsInterface
+     */
+    function supportsInterface(bytes4 interfaceId) external pure returns (bool) {
+        return interfaceId == type(IL2EjectionController).interfaceId || interfaceId == type(ITokenObserver).interfaceId || interfaceId == type(IERC1155Receiver).interfaceId;
+    }
+
+    /**
+     * Implements ERC1155Receiver.onERC1155Received
+     */
+    function onERC1155Received(address, address, uint256, uint256, bytes calldata) external pure returns (bytes4) {
+        return this.onERC1155Received.selector;
+    }
+
+    /**
+     * Implements ERC1155Receiver.onERC1155BatchReceived
+     */
+    function onERC1155BatchReceived(address, address, uint256[] memory, uint256[] memory, bytes calldata) external pure returns (bytes4) {
+        return this.onERC1155BatchReceived.selector;
+    }
+
+    /**
+     * Implements ITokenObserver.onRenew
+     */
+    function onRenew(uint256 tokenId, uint64 expires, address renewedBy) external {
+        if (registry.ownerOf(tokenId) == address(this)) {
+            // this will get picked up by the bridge
+            emit NameRenewed(tokenId, expires, renewedBy);
```

Comment:
Likewise here - this should be in a subclass.

Reply:
But don't we want this base class to atleast emit an event?